### PR TITLE
Move `str` to libcore.

### PR DESCRIPTION
**Pull request by eddyb**
I was reminded we tried to do this before and didn't go through with it for some reason, after seeing https://github.com/rust-lang/unsafe-code-guidelines/issues/78 (the impact on this PR on that question isn't significant, FWIW).

One thing I didn't do is change how `str` is resolved: it's still a builtin type for `rustc_resolve`, which *only after* name resolution is turned into the right type, using the `"str"` lang item.
(Presumably if `core::str` had `crate mod sealed { pub struct Str([u8]); }` then `core::prelude` could do `pub use crate::str::sealed::Str as str;`, but I haven't tried it - IMO a requirement is that users can't observe `core::str::str` or `core::str::Str`. cc @petrochenkov)

While this does seem to work with enough tweaks, I don't have the bandwidth to push this through.
So I suppose this could be considered insta-abandonware :recycle:.

r? @ghost cc @rust-lang/libs @rust-lang/compiler

### CODE REVIEW

**Comment by SimonSapin** [comment]
>  I don't have the bandwidth to push this through.

Simplifying the language by pushing more of it into the standard library sounds nice from a theoretical point of view, but what is the practical benefit that we would gain by having someone spend the bandwidth to finish this?

File: src/librustc_mir/interpret/validity.rs [hunk 1]
```diff
@@ -715,7 +714,8 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValueVisitor<'mir, 'tcx, M>
         fields: impl Iterator<Item = InterpResult<'tcx, Self::V>>,
     ) -> InterpResult<'tcx> {
         match op.layout.ty.kind {
-            ty::Str => {
+            // FIXME(eddyb) does this belong here?
```
Comment (RalfJung) [inline src/librustc_mir/interpret/validity.rs:717]: Well you could remove that branch and then str would not be special any more in terms of validity. ;)

**Comment by eddyb** [comment]
> but what is the practical benefit that we would gain by having someone spend the bandwidth to finish this?

Not sure there is much to finish, it's more about keeping this working and adjusting it due to review comments, throughout the approval process, whichever teams would decide on it.

But also, wow, this is a slowdown. It *really shouldn't*, I think every single place in the compiler this PR makes slower also has been slowing down user-defined types w/o parameters *this entire time*, so it might be worth digging into some of these and trying to fix them.

**Comment by petrochenkov** [comment]
The previous attempt: https://github.com/rust-lang/rust/pull/19612.

**Comment by oli-obk** [comment]
> But also, wow, this is a slowdown. It really shouldn't, I think every single place in the compiler this PR makes slower also has been slowing down user-defined types w/o parameters this entire time, so it might be worth digging into some of these and trying to fix them.

`const_eval_validated` seems to be one of the major regressions here. of the 10% `coercions-check baseline incremental` 5 are in `const_eval_validated`

File: src/librustc_mir/interpret/eval_context.rs [hunk 2]
```diff
@@ -548,7 +548,7 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 Ok(Some(self.read_size_and_align_from_vtable(vtable)?))
             }
 
-            ty::Slice(_) | ty::Str => {
```
Comment (oli-obk) [inline src/librustc_mir/interpret/eval_context.rs:551]: potential site for the perf regression, easy to test by doing the old logic if `ty.is_str()`

File: src/librustc_mir/interpret/operand.rs [hunk 3]
```diff
@@ -314,8 +314,9 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
     // Turn the wide MPlace into a string (must already be dereferenced!)
     pub fn read_str(&self, mplace: MPlaceTy<'tcx, M::PointerTag>) -> InterpResult<'tcx, &str> {
-        let len = mplace.len(self)?;
-        let bytes = self.memory.read_bytes(mplace.ptr, Size::from_bytes(len))?;
+        let bytes_mplace = self.mplace_field(mplace, 0)?;
```
Comment (oli-obk) [inline src/librustc_mir/interpret/operand.rs:317]: potential site for the perf regression. We should probably check whether we can make `mplace_field` a no-op in case of `repr(Rust)` with one field or `repr(transparent)` types.

I think the main problem here is that we compute the field's layout even for cases like this one where we don't actually need the layout. We've seen major regressions in miri due to a single additional call to `tcx.layout_of` before.

**Comment by petrochenkov** [comment]
Note: the diff is actually negative if you ignore the newly added tests.

**Comment by petrochenkov** [comment]
Reviewed.
I still think this brings more benefits than disadvantages.
The PR reminded me my old [de-boxification PR](https://github.com/rust-lang/rust/pull/39230) a lot.

It feels like most of the special treatment in diagnostics can be removed if we officially recognize that `str` is a struct in the standard library + rename `sealed::Str` into `sealed::str`.

The special case in mangling can be removed as well, except that it would needs a version bump.

**Comment by petrochenkov** [comment]
>But also, wow, this is a slowdown. It *really shouldn't*, I think every single place in the compiler this PR makes slower also has been slowing down user-defined types w/o parameters *this entire time*, so it might be worth digging into some of these and trying to fix them.

:+1: 
Worth doing even if the PR itself doesn't land.

**Comment by eddyb** [comment]
> It feels like most of the special treatment in diagnostics can be removed if we officially recognize that str is a struct in the standard library + rename `sealed::Str` into `sealed::str`.

I haven't done any testing on this but I wonder if:
* exporting it only through the prelude would result in `std::prelude::v1::str` paths
* we should enshrine the implicit prelude further and just print e.g. `Vec<String>` and `&str`
  * maybe an opt-out could be the crate defining *or importing* something with the same name as one of the prelude exports, then we would exclude *that* name from the "nice" behavior
  * this is effectively a crate-wide approximation of #21934, for prelude exports

> The special case in mangling can be removed as well, except that it would needs a version bump.

We sort of had plans for that, along the lines of representing paths like `foo::<"abc">` (i.e. `const` generics of type `&'static str`), so presumably we would've reused the same "primitive type" identifier (`e`) that describes `str`, in the encoding of the constant that would contain the string bytes.

However, I realize now that we could encode it as `foo::<&str(*b"abc")>` all the same (and `*b"abc"` would likely be embedded as the hex sequence `616263` anyway, 

We shouldn't need a bump to make the `str` (`e`) "primitive type" obsolete, we would simply not be emitting it, and demanglers would just never see it. We'd only need a bump if we want to *repurpose* that "primitive type" identifier (`e`) for something else.

**Comment by eddyb** [comment]
> * we should enshrine the implicit prelude further and just print e.g. `Vec<String>` and `&str`
>   * maybe an opt-out could be the crate defining *or importing* something with the same name as one of the prelude exports, then we would exclude *that* name from the "nice" behavior
>   * this is effectively a crate-wide approximation of #21934, for prelude exports

To expand on this a bit: one of the `rustc_resolve` outputs could be:
```rust
unshadowed_prelude_items: FxHashMap<DefId, Symbol>,
```
that we use in `try_print_visible_def_path` to show *just* that name and nothing else.

**Comment by petrochenkov** [comment]
>To expand on this a bit: one of the `rustc_resolve` outputs could be:

Yes, `Resolver::prelude` can be converted into such map in `fn into_outputs(self) -> ResolverOutputs`.
Can be done independently from this PR.

>exporting it only through the prelude would result in `std::prelude::v1::str` paths

It would, but that's not a big deal?
We have things like e.g. `std::prelude::v1::include` exposed through the prelude module because built-in macros like `include!()` are introduced through the prelude.
If we ever decide to revert this PR and decide that `str` is built-in after all, we can always keep a single `std::prelude::v1::str` reexport in place for compatibility.

**Comment by eddyb** [comment]
> Yes, `Resolver::prelude` can be converted into such map in `fn into_outputs(self) -> ResolverOutputs`.

Neat! Presumably scanning for locally defined items with the same names would be more effort, but maybe there's some map (oh, there's the `DefPathTable`, that might be an easy way to prototype this).

> EDIT: Unless you were talking about error messages here, `std::prelude::v1::str` certainly shouldn't appear in them.

Sorry, yeah, my concerns are:
* good error messages (*ideally* it stays `str`)
* `std::str::str` or similar doesn't start working instantly on stable (*even* if we eventually decide to export it like that)

**Comment by Centril** [comment]
Thoughts from the language team meeting yesterday:

- @eddyb noted that if the interfaces are all the same, this should be fine.
- We understood the motivation as:
    - Removing special cases, but we will still require special cases for name resolution.
    - Some diagnostics are somewhat better, some worse (maybe those can be fixed).
    - But at almost every other place that referenced `str` before, some kind of `if` is currently required in the compiler, at least in the first implementation.
- @Centril was not sure if it was an improvement, but though that as long as we’re not “locked in” one way or the other, it seems fine.
   - It was noted that we'll still have language items for any additional behavior that the language team might want to add, if we want to add it.
- @eddyb noted that @petrochenkov mentioned that there was an attempt long ago, they’d like to see it done.

So in conclusion, I think we can green-light this from a lang team POV (regressions in perf etc. are for the compiler team to assess, and fix), but let's also be careful not to lock us into a world where this cannot be reverted.

**Comment by petrochenkov** [comment]
Closing due to inactivity.

**Comment by eddyb** [comment]
It's possible #73996 unblocked the "diagnostics should still say `str`, not show a `core::...` path" part of this, but I don't think if I will get to try and revive this any time soon.
